### DISCUSSION TITLE
Request: Connect one remote to multiple hubs

### DISCUSSION THREAD

**Discussion by Stuckalf** [pybricks/support#1463] [Ideas]
Hi all,

I know this has been asked before... but I think it would be a great feature to get a connection between one remote and different hubs (on different BT channels).

I'd guess quite a few Powered Up users are using the system to control their trains. And it is a nice feature (of the original Lego firmware) to switch between five different hubs - all connected to one remote. Even ten hubs are possible if you pair them within the five available channels that the Lego system offers - so you could control up to ten trains with one single remote. 
Controlling ten trains at the same time might not be the case for a normal train scenario. But two or three trains can be a typical usecase, even in a home environment. And I'm talking about different trains, not a single locomotive with multiple engines.

I couldn't figure out how the Lego firmware solves the problem. There is some kind of broadcast going on in the Powered Up system. But if multiple hubs can establish a connection to one single peripheral, each on a dedicated channel, a broadcast on an advertising channel would be unneccessary for this usecase.
Afaik connecting one peripheral (the remote) to multiple central devices is available since BT BLE 4.1. And the Lego developers seem to have done quite a good job on this topic. 

Greetings, Alf

**Comment by laurensvalk** [comment]
Thanks! See this post for an example to do it: https://github.com/pybricks/pybricks-projects/issues/96

Since it's wireless, you can use this for a multi-engine train or for any number of trains.

I guess it amounts to putting up a few good examples on the site. Besides the multi-engine case, what is your preferred use case?

I just ordered a few train motors, so I hope to be making a few more examples.

**Comment by Stuckalf** [comment]
Thanks Laurens, but I think  you've got me wrong.
I'm not searching for a multi-engine option for one locomotive. I'm thinking about switch between whole trainsets or locomotives - which indeed is possible with the original Lego firmware. Each hub connects to the remote on a different BT channel.
It would be nice to get the mail train up and running and - by switching to another hub e.g. locomotive - do the shunting job in the railyard.

In the DCC ecosystem you sometimes use one controller for a bunch of locomotives - so this is a normal way on controlling multiple trains.

**Reply by laurensvalk** [reply]
Sure, all this can be done with Pybricks. I'm just trying to figure out which would be the most useful examples to add to the site :)

It's wireless so as mentioned above it could work for any number of trains. Would you like to toggle through multiple trains using the green button? Or something else? What should the other buttons do, left and right? Control one train? Speed increments?

Thanks!

**Reply by Stuckalf** [reply]
I'd prefer the same logic which can be used in the Lego Powered Up system:
- switch between hubs with the green button
- standard controls for the three left and right buttons (speed and stop)
- possibility to connect two hubs on one channel - so you could connect one train motor with Port A on one hub and another train motor with Port B on the second hub - that would give us 10 trains to control (as mentioned in my post above) and you could control two trains at once without having to switch a channel

**Reply by laurensvalk** [reply]
> standard controls for the three left and right buttons (speed and stop)

Can you elaborate? So basically the same as https://github.com/pybricks/pybricks-projects/issues/96, but more trains?

**Reply by Stuckalf** [reply]
Currently I have no usecase for multiple motors in one single train (if I read project #96 correctly), although this might change in the future.

The original Lego firmware gives us the possibility to switch between multiple hubs when pressing the green center button.
So the most important thing on my wishlist would be to use up to five different Bluetooth channels with one remote - switchable through the green button. And it shouldn't matter if there's one or multiple hubs connected on a single channel. 

I tried to connect two Pybrick hubs with one remote but as soon as I have connected one hub, it seems the advertising of the remote has stopped - the second Pybrick isn't able to recognize the remote so it's still only one hub that can connect to one remote.

And starting with BLE 4.1 it should be possible to connect one peripheral to multiple hubs. Maybe I'm wrong, but I don't think using a hub to broadcast the remote data to other hubs is the wrong way of connecting the hubs. 
I must admit my knowledge of Bluetooth connection is quite superficial.
Thanks for your time and efforts!

**Reply by laurensvalk** [reply]
I'd still like to know what exactly you mean with left and right buttons for control :smile: Does it match the project I linked to?

What I mean is that Pybricks is more like a general purpose toolbox. There isn't really a limit to what you can do, but we wouldn't want to make the main examples overly complex.

In other words, it's a trade off between making it simple to configure in each program, or trying to put every possible use case into a single program so you'd never have to change it. (There is also a portion of users who prefer it this way)

**Reply by Stuckalf** [reply]
No problem 😄
Left and right buttons are the +, - and red center buttons each on the left and right of the remote.
I'd prefer to tap the + and - buttons to increase or decrease the motor speed - like the original Lego firmware does. I think that's the best way to understand the control, because new users know this type of control from the original firmware.

Anyway, I'm not searching for an example how to control a motor.  My problem is that with Pybricks installed on two of my hubs I cannot connect those two hubs to one remote. I couldn't find any Pybrick attribute or option to set the Bluetooth channel for a connection between a hub and a remote - or any means of communication id, so that it's possible to get remote commands only on specific hub.

**Reply by Stuckalf** [reply]
Maybe there's an option to read the current color value of the remote LED. So the script inside the hub could check if the selected "channel" on the remote (shown as color coding) corresponds with the color of the hubs own LED. This would be a kind of "id" to check if a button press on the remote is intended for a certain hub with the same LED color setting. Anything like this would be a good option for the case, that different channels aren't available.

**Reply by cimpeanioan** [reply]
I'm looking for something similar, to send a command from a remote control to 2 or more hubs (like a sync'ed trigger/master switch-on). It's possible with hub-to-hub communication, as I saw in the example of the PyBricks projects but I'd like to use the remote control instead of a hub.

**Reply by laurensvalk** [reply]
@cimpeanioan Here is an example of using one remote to control two hubs. https://pybricks.com/project/technic-42146-powered-up-remote/ The actual connection is still to one hub, which then passes it on to the other hub.

**Reply by cimpeanioan** [reply]
@laurensvalk thank you for replying! 
I ended up using the remote just like you mentioned (already looked over the crane project):

- remote <-> central hub
- central hub -> broadcasting to multiple "auxiliary" hubs

The delay is not that noticeable. So far so good!

![Remote + Hub1 -  Hub2, Hub3](https://github.com/user-attachments/assets/374772aa-cb92-4c24-96a4-9c3ef87ec638)

**Reply by laurensvalk** [reply]
Very nice!
